src/topology/semicontinuous.lean

```diff
@@ -0,0 +1,833 @@
+/-
+Copyright (c) 2021 Sébastien Gouëzel. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sébastien Gouëzel
+-/
+import topology.continuous_on
+import data.indicator_function
+import topology.algebra.group
+import topology.algebra.ordered.liminf_limsup
+import topology.instances.ennreal
+
+/-!
+# Semicontinuous maps
+
+A function `f` from a topological space `α` to an ordered space `β` is lower semicontinuous at a
+point `x` if, for any `ε > 0`, for any `x'` close enough to `x`, one has `f x' > f x - ε`. In other
+words, `f` can jump up, but it can not jump down. Upper semicontinuous functions are defined
+similarly.
+
+This file introduces these notions, and a basic API around them mimicking the API for continuous
+functions.
+
+## Main definitions and results
+
+We introduce 4 definitions related to lower semicontinuity:
+* `lower_semicontinuous_within_at f s x`
+* `lower_semicontinuous_at f x`
+* `lower_semicontinuous_on f s`
+* `lower_semicontinuous f`
+
+We build a basic API using dot notation around these notions, and we prove that
+* constant functions are lower semicontinuous;
+* `indicator s (λ _, y)` is lower semicontinuous when `s` is open and `0 ≤ y`, or when `s` is closed
+  and `y ≤ 0`;
+* continuous functions are lower semicontinuous;
+* a sum of two (or finitely many) lower semicontinuous functions is lower semicontinuous;
+* a supremum of a family of lower semicontinuous functions is lower semicontinuous;
+* An infinite sum of `ℝ≥0∞`-valued lower semicontinuous functions is lower semicontinuous.
+
+Similar results are stated and proved for upper semicontinuity.
+
+We also prove that a function is continuous if and only if it is both lower and upper
+semicontinuous.
+
+## Implementation details
+
+All the nontrivial results for upper semicontinuous functions are deduced from the corresponding
+ones for lower semicontinuous functions using `order_dual`.
+
+-/
+
+open_locale topological_space big_operators ennreal
+open set
+
+variables {α : Type*} [topological_space α] {β : Type*} [preorder β]
+{f g : α → β} {x : α} {s t : set α} {y z : β}
+
+/-!
+### Lower semicontinuous functions
+-/
+
+/-- A function `f` is lower semicontinuous at `x` within a set `s` if, for any `ε > 0`, for all
+`x'` close enough to `x` in  `s`, then `f x'` is at least `f x - ε`. -/
+def lower_semicontinuous_within_at (f : α → β) (s : set α) (x : α) :=
+∀ y < f x, ∀ᶠ x' in 𝓝[s] x, y < f x'
```

Comment:
I wonder if the "more familiar" statement you give in the doc-string obscures the actual definition you give here, which in particular doesn't require any subtraction operation!

I appreciate the desire to mention the standard definition...

Reply:
I have tried to clarify the docstrings. Let me know if you like it better. Otherwise, I can even remove all mention of \epsilon.